### PR TITLE
fix(notebook-controller): correct HTTPRoute port for unauthenticated notebooks

### PR DESCRIPTION
**Pull request by harshad16**
# Fix: HTTPRoute port mismatch for unauthenticated notebooks

**Fixes:** [RHOAIENG-39253](https://redhat.atlassian.net/browse/RHOAIENG-39253)

## Summary

Unauthenticated notebooks (without `inject-auth` annotation) were completely inaccessible due to HTTPRoute referencing port 8888 instead of the Service's actual port 80. This fix changes the HTTPRoute configuration to use the correct Service port per Gateway API specification.

## Problem

When a notebook is created **without** the `notebooks.opendatahub.io/inject-auth` annotation:
- ODH controller creates HTTPRoute pointing to port **8888**
- Upstream Kubeflow controller creates Service exposing port **80** (targetPort: 8888)
- Gateway API `BackendRef.Port` must reference the **Service port**, not the container port
- Result: Gateway cannot resolve backend → **503 Service Unavailable**
- **Impact:** 100% of unauthenticated notebooks are unreachable

## Root Cause

Bug introduced in [RHOAIENG-34310](https://redhat.atlassian.net/browse/RHOAIENG-34310) (commit bf6b20b5a, 2025-10-16) when implementing Gateway API HTTPRoute. The port mismatch existed from inception but wasn't caught because:
- Tests validated the incorrect behavior (expected port 8888)
- No E2E tests for unauthenticated notebook path
- All testing focused on authenticated notebooks

## Solution

Change HTTPRoute backend port from **8888** to **80** to align with Service configuration.

**Files Changed:**
1. `notebook_route.go:120-122` - HTTPRoute port configuration
2. `notebook_route.go:302-303` - Route detection for cleanup logic
3. `notebook_controller_test.go` - 5 test assertions updated

**Before:**
```go
Port: (*gatewayv1.PortNumber)(&[]gatewayv1.PortNumber{8888}[0])  // Wrong - container port
```

**After:**
```go
// Fix for RHOAIENG-39253: Use Service port (80), not container port (8888)
// Gateway API BackendRef.Port must reference the Service's exposed port
Port: (*gatewayv1.PortNumber)(&[]gatewayv1.PortNumber{80}[0])
```

**Why This Works:**
- Gateway API routing: `Client → Gateway → HTTPRoute (Service Port) → Service (Port) → Pod (TargetPort)`
- Correct flow: `HTTPRoute:80 → Service:80 → Pod:8888` ✅
- Previous flow: `HTTPRoute:8888 → Service:8888 (doesn't exist) → FAIL` ❌

## Testing

**✅ Completed:**
- Code compiles successfully
- Basic unit tests pass (12/12)
- Test assertions updated to expect correct behavior (port 80)

**⚠️ Required Before Merge:**
- **Integration tests must pass in CI/CD** (blocked locally - missing kubebuilder/envtest)
- 107 Ginkgo specs must all pass
- **Do not merge if integration tests fail**

**Recommended:**
- Manual verification in test cluster
- Test mode switching (add/remove inject-auth annotation)
- Verify migration scenarios (legacy inject-oauth annotation)

### Automated Testing
```bash
# Run integration tests (requires kubebuilder/envtest)
make envtest
go test -v ./controllers/...
# All tests must pass
```

### Manual Testing (Recommended)
```bash
# 1. Create unauthenticated notebook
cat <<EOF | kubectl apply -f -
apiVersion: kubeflow.org/v1
kind: Notebook
metadata:
  name: test-no-auth
  namespace: test-namespace
spec:
  template:
    spec:
      containers:
      - name: notebook
        image: quay.io/opendatahub/notebooks:jupyter-minimal
EOF

# 2. Verify HTTPRoute uses port 80
kubectl get httproute -n <controller-namespace> nb-test-namespace-test-no-auth -o yaml
# Check: spec.rules[0].backendRefs[0].port == 80

# 3. Verify Service exposes port 80
kubectl get svc -n test-namespace test-no-auth -o yaml
# Check: spec.ports[0].port == 80, targetPort == 8888

# 4. Test connectivity
curl -k https://<gateway-url>/notebook/test-namespace/test-no-auth
# Expected: Jupyter UI (not 503)

# 5. Test mode switching
kubectl annotate notebook test-no-auth notebooks.opendatahub.io/inject-auth=true
# Verify: old HTTPRoute deleted, new kube-rbac-proxy HTTPRoute created

kubectl annotate notebook test-no-auth notebooks.opendatahub.io/inject-auth-
# Verify: kube-rbac-proxy HTTPRoute deleted, regular HTTPRoute (port 80) recreated
```

## Deployment Impact

**Severity:** High - Fixes completely broken functionality

**Affected Users:**
- Anyone using notebooks without auth-proxy
- Deployments opting out of authentication
- Migration from RHOAI 2.x with legacy annotations

**Breaking Changes:** None

**Migration Required:** None
- Existing notebooks will automatically work after controller restart
- HTTPRoutes reconciled with correct port

**Rollback:** Safe
- Reverting this change returns to broken state (notebooks inaccessible)
- No data loss or corruption risk

## Follow-Up Work

1. **High Priority:** Add E2E test for unauthenticated notebook connectivity
   - Issue to be created post-merge
   - Prevents regression of this exact bug

2. **Medium Priority:** Test migration scenarios in staging
   - Legacy `inject-oauth` annotation handling
   - Upgrade path validation

3. **Low Priority:** Investigate other missing resources
   - Per AI triage comment, verify ServiceAccount/NetworkPolicy requirements

## Comparison: Authenticated vs Unauthenticated

**Authenticated notebooks (working correctly):**
- HTTPRoute port: 8443
- Service (`<notebook>-kube-rbac-proxy`) port: 8443
- Match → routing works ✅

**Unauthenticated notebooks (before fix):**
- HTTPRoute port: 8888
- Service (`<notebook>`) port: 80
- Mismatch → routing fails ❌

**Unauthenticated notebooks (after fix):**
- HTTPRoute port: 80
- Service (`<notebook>`) port: 80
- Match → routing works ✅

## Related Issues

- [RHOAIENG-34310](https://redhat.atlassian.net/browse/RHOAIENG-34310) - Introduced Gateway API HTTPRoute (where bug originated)
- [RHOAIENG-38009](https://redhat.atlassian.net/browse/RHOAIENG-38009) - Central namespace refactor (didn't address port issue)

## Checklist

- [x] Code follows project conventions
- [x] Comments explain WHY, not just WHAT
- [x] Issue number referenced in comments
- [x] Tests updated to reflect correct behavior
- [x] Code compiles without errors
- [x] Basic unit tests pass
- [ ] **Integration tests pass** (CI/CD required)
- [ ] Manual verification recommended
- [x] Documentation artifacts created
- [x] No breaking changes
- [x] No security vulnerabilities introduced

## Review Focus Areas

1. **Port number correctness** - Verify 80 is the right Service port
2. **Cleanup logic** - Ensure route detection works for port 80
3. **Test coverage** - Check integration tests pass in CI/CD
4. **Edge cases** - Migration scenarios, concurrent reconciliation

## Co-Authored-By

This fix was implemented with assistance from Claude (AI pair programmer).

---

**Ready for Review** ✅
**Merge Status:** ⚠️ Pending integration tests in CI/CD

## Release Notes

* **Bug Fixes**
  * Updated notebook traffic routing configuration to use Service port 80 instead of container port 8888 in Gateway API HTTPRoute backend references.

[RHOAIENG-39253]: https://redhat.atlassian.net/browse/RHOAIENG-39253?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

[RHOAIENG-34310]: https://redhat.atlassian.net/browse/RHOAIENG-34310?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ
[RHOAIENG-34310]: https://redhat.atlassian.net/browse/RHOAIENG-34310?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

File: components/odh-notebook-controller/controllers/notebook_controller_test.go [hunk 1]
```diff
@@ -104,7 +104,8 @@ var _ = Describe("The Openshift Notebook controller", func() {
 										Kind:      func() *gatewayv1.Kind { k := gatewayv1.Kind("Service"); return &k }(),
 										Name:      gatewayv1.ObjectName(Name),
 										Namespace: func() *gatewayv1.Namespace { ns := gatewayv1.Namespace(Namespace); return &ns }(), // Cross-namespace reference
-										Port:      (*gatewayv1.PortNumber)(&[]gatewayv1.PortNumber{8888}[0]),
+										// Fix for RHOAIENG-39253: Use Service port (80)
+										Port: (*gatewayv1.PortNumber)(&[]gatewayv1.PortNumber{80}[0]),
```
[the remark this answers is not in the document]
Comment (jstourac) [reply components/odh-notebook-controller/controllers/notebook_controller_test.go:108]: This will be handled via #794.

File: components/odh-notebook-controller/controllers/notebook_controller_test.go [hunk 2]
```diff
@@ -587,7 +588,8 @@ var _ = Describe("The Openshift Notebook controller", func() {
 										Kind:      func() *gatewayv1.Kind { k := gatewayv1.Kind("Service"); return &k }(),
 										Name:      gatewayv1.ObjectName(Name),
 										Namespace: func() *gatewayv1.Namespace { ns := gatewayv1.Namespace(Namespace); return &ns }(),
-										Port:      (*gatewayv1.PortNumber)(&[]gatewayv1.PortNumber{8888}[0]),
+										// Fix for RHOAIENG-39253: Use Service port (80)
+										Port: (*gatewayv1.PortNumber)(&[]gatewayv1.PortNumber{80}[0]),
```
[the remark this answers is not in the document]
Comment (jstourac) [reply components/odh-notebook-controller/controllers/notebook_controller_test.go:611]: This will be handled via #794.

File: components/odh-notebook-controller/controllers/notebook_controller_test.go [hunk 3]
```diff
@@ -1525,7 +1529,8 @@ var _ = Describe("The Openshift Notebook controller", func() {
 			Expect(len(httpRoute.Spec.Rules)).To(BeNumerically(">", 0))
 			Expect(len(httpRoute.Spec.Rules[0].BackendRefs)).To(BeNumerically(">", 0))
 			Expect(string(httpRoute.Spec.Rules[0].BackendRefs[0].BackendRef.BackendObjectReference.Name)).To(Equal(Name))
-			Expect(*httpRoute.Spec.Rules[0].BackendRefs[0].BackendRef.BackendObjectReference.Port).To(Equal(gatewayv1.PortNumber(8888)))
+			// Fix for RHOAIENG-39253: Unauthenticated HTTPRoute should use Service port (80)
+			Expect(*httpRoute.Spec.Rules[0].BackendRefs[0].BackendRef.BackendObjectReference.Port).To(Equal(gatewayv1.PortNumber(80)))
```
[the remark this answers is not in the document]
Comment (jstourac) [reply components/odh-notebook-controller/controllers/notebook_controller_test.go:1571]: This will be handled via #794.

File: components/odh-notebook-controller/controllers/notebook_route.go [hunk 4]
```diff
@@ -117,7 +117,9 @@ func NewNotebookHTTPRoute(notebook *nbv1.Notebook, centralNamespace string) *gat
 								BackendObjectReference: gatewayv1.BackendObjectReference{
 									Name:      gatewayv1.ObjectName(notebook.Name),
 									Namespace: (*gatewayv1.Namespace)(&notebook.Namespace), // Cross-namespace reference
-									Port:      (*gatewayv1.PortNumber)(&[]gatewayv1.PortNumber{8888}[0]),
+									// Fix for RHOAIENG-39253: Use Service port (80), not container port (8888)
+									// Gateway API BackendRef.Port must reference the Service's exposed port
+									Port: (*gatewayv1.PortNumber)(&[]gatewayv1.PortNumber{80}[0]),
```
[the remark this answers is not in the document]
Comment (jstourac) [reply components/odh-notebook-controller/controllers/notebook_route.go:122]: This will be handled via #794.

**Comment by harshad16** [comment]
Closing this one as it is stale and would be re-looked in future
Thanks for all the reviews
